File: host.py
import socket
import json
import threading
import time
import struct
import sys
import os

# =========================================================================
#   CẤU HÌNH TRỰC TIẾP TẠI ĐÂY (KHÔNG CẦN FILE CONFIG)
# =========================================================================
SERVER_HOST = "mbasic7.pikamc.vn"
SERVER_PORT = 25284
CLIENT_ID   = "bedrock-host-1"
LOCAL_HOST  = "169.254.238.23" # IP LAN của bạn
LOCAL_PORT  = 7551            # Port Minecraft đang mở (7551 hoặc 19132)
MY_KEY      = ""               # Để trống để tự sinh key mới
AUTH_TOKEN  = ""
# =========================================================================

class UDP_MSG:
    HANDSHAKE = 0x01
    DATA      = 0x02
    CLOSE     = 0x03
    PING      = 0x04
    PONG      = 0x05

def build_udp_message(msg_type, key, session_id, payload=None):
    if payload is None: payload = b''
    k_bytes = key.encode('utf8')
    s_bytes = session_id.encode('utf8')
    header = struct.pack('>B H', msg_type, len(k_bytes)) + k_bytes + struct.pack('>H', len(s_bytes)) + s_bytes
    return header + payload

def parse_udp_message(data):
    try:
        msg_type = data[0]
        k_len = struct.unpack('>H', data[1:3])[0]
        key = data[3:3+k_len].decode('utf8')
        s_len = struct.unpack('>H', data[3+k_len:5+k_len])[0]
        sid = data[5+k_len:5+k_len+s_len].decode('utf8')
        payload = data[5+k_len+s_len:]
        return {"type": msg_type, "key": key, "id": sid, "payload": payload}
    except: return None

class BedrockAgent:
    def __init__(self):
        self.server_host = SERVER_HOST
        self.server_port = SERVER_PORT
        self.client_id = CLIENT_ID
        self.local_host = LOCAL_HOST
        self.local_port = LOCAL_PORT
        self.my_key = MY_KEY
        self.auth_token = AUTH_TOKEN
        
        self.udp_ready = False
        self.udp_sock = None
        self.sessions = {} # sid -> socket
        self.running = True

    def log(self, tag, msg, color='\033[0m'):
        print(f"{color}[{tag}] {msg}\033[0m")

    def connect_control(self):
        while self.running:
            try:
                self.log("Control", f"Connecting to {self.server_host}:{self.server_port}...", '\033[36m')
                tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                tcp_sock.settimeout(10)
                tcp_sock.connect((self.server_host, self.server_port))
                tcp_sock.settimeout(None)
                
                # Register
                reg = {
                    "type": "register",
                    "key": self.my_key,
                    "token": self.auth_token,
                    "client_id": self.client_id,
                    "target": f"{self.local_host}:{self.local_port}",
                    "protocol": "udp"
                }
                tcp_sock.sendall((json.dumps(reg) + "\n").encode('utf8'))
                
                # Read loop
                buf = ""
                while self.running:
                    raw = tcp_sock.recv(4096)
                    if not raw: break
                    buf += raw.decode('utf8', errors='ignore')
                    while "\n" in buf:
                        if "\n" not in buf: break
                        line, buf = buf.split("\n", 1)
                        if line.strip():
                            try:
                                msg = json.loads(line)
                                self.handle_control(msg, tcp_sock)
                            except: pass
                
                tcp_sock.close()
            except Exception as e:
                self.log("Error", f"Control failed: {e}", '\033[31m')
            
            self.udp_ready = False
            time.sleep(3)

    def start(self):
        self.connect_control()

    def handle_control(self, msg, tcp_sock):
        mtype = msg.get('type')
        if mtype == 'registered':
            self.my_key = msg.get('key', self.my_key)
            self.log("OK", f"Registered! Key: {self.my_key[:8]}", '\033[32m')
            threading.Thread(target=self.setup_udp_data, daemon=True).start()
            threading.Thread(target=self.tcp_ping_loop, args=(tcp_sock,), daemon=True).start()
        elif mtype == 'udp_open':
            self.open_udp_session(msg.get('id'))
        elif mtype == 'udp_close':
            self.close_udp_session(msg.get('id'))
        elif mtype == 'ping':
            try: tcp_sock.sendall(json.dumps({"type": "pong"}).encode('utf8') + b"\n")
            except: pass
        elif mtype == 'error':
            self.log("Server Error", msg.get('error'), '\033[31m')

    def tcp_ping_loop(self, sock):
        while self.running:
            try:
                time.sleep(15)
                sock.sendall(json.dumps({"type": "pong"}).encode('utf8') + b"\n")
            except: break

    def setup_udp_data(self):
        self.udp_ready = False
        try:
            self.udp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            def handshake():
                while not self.udp_ready and self.running:
                    pkt = build_udp_message(UDP_MSG.HANDSHAKE, self.my_key, "")
                    self.udp_sock.sendto(pkt, (self.server_host, self.server_port))
                    time.sleep(1)
            threading.Thread(target=handshake, daemon=True).start()
            
            while self.running:
                data, addr = self.udp_sock.recvfrom(65535)
                parsed = parse_udp_message(data)
                if not parsed or parsed['key'] != self.my_key: continue
                
                ptype = parsed['type']
                if ptype == UDP_MSG.HANDSHAKE:
                    if not self.udp_ready:
                        self.udp_ready = True
                        self.log("UDP", "Data channel connected!", '\033[32m')
                elif ptype == UDP_MSG.DATA:
                    sid = parsed['id']
                    if sid in self.sessions:
                        self.sessions[sid].sendto(parsed['payload'], (self.local_host, self.local_port))
                elif ptype == UDP_MSG.PING:
                    pong = build_udp_message(UDP_MSG.PONG, self.my_key, parsed['id'], parsed['payload'])
                    self.udp_sock.sendto(pong, (self.server_host, self.server_port))
        except: pass

    def open_udp_session(self, sid):
        if sid in self.sessions: self.close_udp_session(sid)
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.bind(('0.0.0.0', 0))
            self.sessions[sid] = sock
            self.log("Link", f"Player joined. Session: {sid}", '\033[34m')
            def game_to_server():
                while sid in self.sessions and self.running:
                    try:
                        data, addr = sock.recvfrom(65535)
                        if self.udp_ready:
                            pkt = build_udp_message(UDP_MSG.DATA, self.my_key, sid, data)
                            self.udp_sock.sendto(pkt, (self.server_host, self.server_port))
                    except: break
            threading.Thread(target=game_to_server, daemon=True).start()
        except: pass

    def close_udp_session(self, sid):
        if sid in self.sessions:
            try: self.sessions[sid].close()
            except: pass
            del self.sessions[sid]
            self.log("End", f"Player left. Session: {sid}", '\033[31m')

if __name__ == "__main__":
    print(f"\n\033[35m=== BEDROCK TUNNEL AGENT (Python FRP) ===\033[0m")
    print(f"Target: {LOCAL_HOST}:{LOCAL_PORT}")
    print(f"-----------------------------------------\n")
    agent = BedrockAgent()
    try:
        agent.start()
    except KeyboardInterrupt:
        print("\nStopping...")
        agent.running = False
        sys.exit(0)

File: src/host.js
const net = require('net');
const dgram = require('dgram');
const crypto = require('crypto');
const fs = require('fs');
const { createLineParser, encodeControl, buildUDPMessage, parseUDPMessage, UDP_MSG } = require('./common/protocol');

function loadConfig(path) {
  try {
    return JSON.parse(fs.readFileSync(path, 'utf8'));
  } catch (e) {
    process.exit(1);
  }
}

function startAgent(config) {
  const { serverHost, serverControlPort: serverPort, clientId, localHost, localUdpPort: localPort } = config;
  const reconnectMs = config.reconnectMs || 3000;
  let myKey = config.key || '';
  let authenticated = false;
  let udpCtrlConn = null;
  let udpReady = false;
  const udpSessions = new Map();

  // ── UDP Data Relay ───────────────────────────────────────────────────

  function setupUDPChannel() {
    if (udpCtrlConn) try { udpCtrlConn.close(); } catch (_) {}
    udpReady = false;

    // Use a high-performance UDP socket with DNS lookup disabled
    const sock = dgram.createSocket({ type: 'udp4', lookup: (hostname, options, cb) => cb(null, hostname, 4) });
    udpCtrlConn = sock;

    sock.on('message', (buf) => {
      const parsed = parseUDPMessage(buf);
      if (!parsed || parsed.key !== myKey) return;

      if (parsed.msgType === UDP_MSG.HANDSHAKE) {
        if (!udpReady) {
          udpReady = true;
          console.log('\x1b[32m[UDP Handshake]\x1b[0m Tunneling active.');
        }
      } else if (parsed.msgType === UDP_MSG.DATA) {
        const sess = udpSessions.get(parsed.id);
        if (sess && parsed.payload.length > 0) {
          // Relay DATA from Server to Game
          sess.send(parsed.payload, localPort, localHost);
        }
      } else if (parsed.msgType === UDP_MSG.CLOSE) {
        closeUDPSession(parsed.id);
      } else if (parsed.msgType === UDP_MSG.PING) {
        sock.send(buildUDPMessage(UDP_MSG.PONG, myKey, parsed.id, parsed.payload), serverPort, serverHost);
      }
    });

    const trigger = () => {
      if (udpCtrlConn === sock && !udpReady) {
        sock.send(buildUDPMessage(UDP_MSG.HANDSHAKE, myKey, '', null), serverPort, serverHost);
      }
    };
    setInterval(trigger, 1000);
    trigger();
  }

  function handleUDPOpen(msg) {
    if (udpSessions.has(msg.id)) closeUDPSession(msg.id);

    const sock = dgram.createSocket('udp4');
    
    // Relay DATA from Game to Server
    sock.on('message', (buf) => {
      if (udpCtrlConn && udpReady) {
        const pkt = buildUDPMessage(UDP_MSG.DATA, myKey, msg.id, buf);
        udpCtrlConn.send(pkt, serverPort, serverHost);
      }
    });

    sock.on('error', () => closeUDPSession(msg.id));

    sock.bind(0, () => {
      udpSessions.set(msg.id, sock);
      console.log(`\x1b[34m[Link]\x1b[0m Session \x1b[33m${msg.id}\x1b[0m: Player → Game (${localPort})`);
    });
  }

  function closeUDPSession(id) {
    const s = udpSessions.get(id);
    if (s) { try { s.close(); } catch (_) {} udpSessions.delete(id); }
  }

  // ── Control Connection ─────────────────────────────────────────────────

  function connectControl() {
    const socket = net.connect({ host: serverHost, port: serverPort }, () => {
      socket.setNoDelay(true);
      console.log(`\x1b[36m[Control]\x1b[0m Connected.`);
      socket.write(encodeControl({
        type: 'register', key: myKey, token: config.authToken || '', client_id: clientId, target: `localhost:${localPort}`, protocol: 'udp'
      }));
    });

    const parser = createLineParser((msg) => {
      switch (msg.type) {
        case 'registered':
          myKey = msg.key || myKey;
          authenticated = true;
          console.log(`\x1b[32m[Success]\x1b[0m Key: ${myKey.slice(0, 8)}...`);
          setupUDPChannel();
          break;
        case 'udp_open':
          if (authenticated) handleUDPOpen(msg);
          break;
        case 'udp_close':
          closeUDPSession(msg.id);
          break;
        case 'ping':
          socket.write(encodeControl({ type: 'pong' }));
          break;
      }
    }, () => socket.destroy());

    socket.on('data', parser);
    socket.on('close', () => {
      authenticated = false; udpReady = false;
      setTimeout(connectControl, reconnectMs);
    });
    socket.on('error', () => {});
  }

  connectControl();
}

startAgent(loadConfig(process.argv[2] || 'configs/agent.config.json'));
